File: experiments/adrianco/experiment-65-fable51-effort/rest-api-crud/runs/effort=default_language=typescript_model=claude-fable-5-1_prompt=neutral/rep1/src/validation.ts
import type { BookInput } from "./db.js";

export interface FieldError {
  field: string;
  message: string;
}

export type ValidationResult =
  | { ok: true; value: BookInput }
  | { ok: false; errors: FieldError[] };

const MAX_TEXT_LENGTH = 500;
const MIN_YEAR = -3000;
const MAX_YEAR = new Date().getUTCFullYear() + 5;

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

/** Strips hyphens/spaces and upper-cases the check digit, e.g. "0-306-40615-2" -> "0306406152". */
export function normalizeIsbn(raw: string): string {
  return raw.replace(/[-\s]/g, "").toUpperCase();
}

/** Accepts ISBN-10 (9 digits + digit/X) or ISBN-13 (13 digits), hyphens/spaces allowed. */
export function isValidIsbn(raw: string): boolean {
  const s = normalizeIsbn(raw);
  if (/^\d{13}$/.test(s)) {
    const sum = [...s].reduce(
      (acc, ch, i) => acc + Number(ch) * (i % 2 === 0 ? 1 : 3),
      0,
    );
    return sum % 10 === 0;
  }
  if (/^\d{9}[\dX]$/.test(s)) {
    const sum = [...s].reduce((acc, ch, i) => {
      const v = ch === "X" ? 10 : Number(ch);
      return acc + v * (10 - i);
    }, 0);
    return sum % 11 === 0;
  }
  return false;
}

function requiredText(
  body: Record<string, unknown>,
  field: string,
  errors: FieldError[],
): string | undefined {
  const value = body[field];
  if (value === undefined || value === null) {
    errors.push({ field, message: `${field} is required` });
    return undefined;
  }
  if (typeof value !== "string") {
    errors.push({ field, message: `${field} must be a string` });
    return undefined;
  }
  const trimmed = value.trim();
  if (trimmed.length === 0) {
    errors.push({ field, message: `${field} must not be empty` });
    return undefined;
  }
  if (trimmed.length > MAX_TEXT_LENGTH) {
    errors.push({ field, message: `${field} must be at most ${MAX_TEXT_LENGTH} characters` });
    return undefined;
  }
  return trimmed;
}

/**
 * Validates a request body for creating or fully replacing a book.
 * `title` and `author` are required; `year` and `isbn` are optional (null when omitted).
 */
export function validateBookInput(body: unknown): ValidationResult {
  if (!isPlainObject(body)) {
    return { ok: false, errors: [{ field: "body", message: "request body must be a JSON object" }] };
  }

  const errors: FieldError[] = [];
  const title = requiredText(body, "title", errors);
  const author = requiredText(body, "author", errors);

  let year: number | null = null;
  if (body.year !== undefined && body.year !== null) {
    if (typeof body.year !== "number" || !Number.isInteger(body.year)) {
      errors.push({ field: "year", message: "year must be an integer" });
    } else if (body.year < MIN_YEAR || body.year > MAX_YEAR) {
      errors.push({ field: "year", message: `year must be between ${MIN_YEAR} and ${MAX_YEAR}` });
    } else {
      year = body.year;
    }
  }

  let isbn: string | null = null;
  if (body.isbn !== undefined && body.isbn !== null) {
    if (typeof body.isbn !== "string") {
      errors.push({ field: "isbn", message: "isbn must be a string" });
    } else {
      const trimmed = body.isbn.trim();
      if (trimmed.length === 0) {
        errors.push({ field: "isbn", message: "isbn must not be empty" });
      } else if (!isValidIsbn(trimmed)) {
        errors.push({ field: "isbn", message: "isbn must be a valid ISBN-10 or ISBN-13" });
      } else {
        isbn = normalizeIsbn(trimmed);
      }
    }
  }

  if (errors.length > 0 || title === undefined || author === undefined) {
    return { ok: false, errors };
  }
  return { ok: true, value: { title, author, year, isbn } };
}

/** Parses a path parameter as a positive integer ID, or returns undefined. */
export function parseId(raw: string | undefined): number | undefined {
  if (raw === undefined || !/^\d+$/.test(raw)) return undefined;
  const id = Number(raw);
  return Number.isSafeInteger(id) && id > 0 ? id : undefined;
}

File: experiments/adrianco/experiment-74-opus55-effort/rest-api-crud/runs/effort=high_language=python_model=claude-opus-5-5_prompt=neutral/rep1/books_api/validation.py
"""Validation of book payloads."""

import datetime
import re

ALLOWED_FIELDS = {"title", "author", "year", "isbn"}
MAX_TEXT_LENGTH = 500
MIN_YEAR = -5000
_ISBN_RE = re.compile(r"^(\d{9}[\dX]|\d{13})$")


def validate_book(payload):
    """Validate a book payload.

    Returns ``(book, errors)`` where ``book`` is a normalised dict (strings
    stripped, ISBN without hyphens/spaces) and ``errors`` maps field names to
    messages. ``book`` is ``None`` if there are any errors.
    """
    if not isinstance(payload, dict):
        return None, {"body": "request body must be a JSON object"}

    errors = {}
    book = {}

    for field in sorted(set(payload) - ALLOWED_FIELDS):
        errors[field] = "unknown field"

    for field in ("title", "author"):
        value = payload.get(field)
        if value is None:
            errors[field] = f"{field} is required"
        elif not isinstance(value, str):
            errors[field] = f"{field} must be a string"
        elif not value.strip():
            errors[field] = f"{field} must not be empty"
        elif len(value.strip()) > MAX_TEXT_LENGTH:
            errors[field] = f"{field} must be at most {MAX_TEXT_LENGTH} characters"
        else:
            book[field] = value.strip()

    year = payload.get("year")
    if year is not None:
        max_year = datetime.date.today().year + 1
        # bool is a subclass of int; reject it explicitly.
        if isinstance(year, bool) or not isinstance(year, int):
            errors["year"] = "year must be an integer"
        elif not MIN_YEAR <= year <= max_year:
            errors["year"] = f"year must be between {MIN_YEAR} and {max_year}"
        else:
            book["year"] = year
    else:
        book["year"] = None

    isbn = payload.get("isbn")
    if isbn is not None:
        if not isinstance(isbn, str):
            errors["isbn"] = "isbn must be a string"
        else:
            normalised = re.sub(r"[\s-]", "", isbn).upper()
            if not normalised:
                book["isbn"] = None
            elif not _ISBN_RE.match(normalised):
                errors["isbn"] = "isbn must be a 10- or 13-digit ISBN"
            else:
                book["isbn"] = normalised
    else:
        book["isbn"] = None

    if errors:
        return None, errors
    return book, {}
